drivers/sdp3x/sdp3x.c

```diff
@@ -0,0 +1,473 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sdp3x
+ * @{
+ * @file
+ * @brief       Sensirion SDP3x sensor driver implementation
+ *
+ * @author      Dirk Ehmen       <ehmen@ibr.cs.tu-bs.de>
+ * @author      Jan Schlichter   <schlichter@ibr.cs.tu-bs.de>
+ * @author      Nishchay Agrawal <f2016088@pilani.bits-pilani.ac.in>
+ * @}
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include "sdp3x_params.h"
+#include "sdp3x.h"
+#include "checksum/crc8.h"
+#include "thread.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define DEV_I2C      (dev->params.i2c_dev)
+#define DEV_ADDR     (dev->params.i2c_addr)
+#define DEV_MODEL    (dev->params.model)
+
+static int8_t _checkCRC(uint16_t value, uint8_t test);
+static int8_t _SDP3x_read_data(const sdp3x_t *dev, int16_t *data);
+static double _SDP3x_convert_to_pascal(const sdp3x_t *dev, int16_t value);
+static double _SDP3x_convert_to_inches_h2o(const sdp3x_t *dev,
+                                           int16_t value) __attribute__(
+    (unused));
+static double _SDP3x_convert_to_celsius(int16_t value);
+int8_t _SDP3x_start_triggered(const sdp3x_t *dev, uint8_t tempComp,
+                              uint8_t clkStretching);
+double _SDP3x_read_temp(const sdp3x_t *dev);
+double _SDP3x_read_pressure(const sdp3x_t *dev);
+
+int SDP3x_init(sdp3x_t *dev, const sdp3x_params_t *params)
+{
+    DEBUG("[SDP3x] init: Initializing device\n");
+
+    dev->params = *params;
+
+    i2c_init(DEV_I2C);
+    SDP3x_soft_reset(dev);
+
+    /* try to read product number to check if sensor is connected and working */
+    int ret = 0;
+    uint8_t cmd1[2] = { 0x36, 0x7C };
+    uint8_t cmd2[2] = { 0xE1, 0x02 };
+    i2c_acquire(DEV_I2C);
+    ret = i2c_write_bytes(DEV_I2C, DEV_ADDR, cmd1, 2, I2C_NOSTOP);
+    if (ret != 0) {
+        i2c_release(DEV_I2C);
+        return ret;
+    }
+    ret = i2c_write_bytes(DEV_I2C, DEV_ADDR, cmd2, 2, I2C_NOSTOP);
+    if (ret != 0) {
+        i2c_release(DEV_I2C);
+        return ret;
+    }
+    i2c_release(I2C_DEV(0));
+
+#ifdef SDP3X_PARAM_IRQ_PIN
+    gpio_init(params->irq_pin, GPIO_IN);
+#endif
+
+    DEBUG("[SDP3x] init: Init done\n");
+    return 1;
+}
+
+double SDP3x_read_single_temperature(const sdp3x_t *dev, uint8_t tempComp,
+                                     uint8_t clkStretching)
+{
+    _SDP3x_start_triggered(dev, tempComp, clkStretching);
+#ifndef SDP3X_PARAM_IRQ_PIN
+    /* Wait for measurement to be ready if irq pin not used */
+    xtimer_usleep(46000);
+#else
+    uint32_t time_now = xtimer_now_usec();
+    /* If for some reason the pin values are not as expected */
+    if (get_irq_pin_value(dev) == 0) {
+        xtimer_usleep(46000);
+    }
+    /* Check for change in IRQn pin to signal data is ready */
+    else {
+        while (get_irq_pin_value(dev) != 0 &&
+               (xtimer_now_usec() - time_now) < 46000) {}
+    }
```

Comment:
If you have an irq pin, I would expect you to actually use irqs instead of polling :wink: 
There is [`gpio_init_int()`](https://doc.riot-os.org/group__drivers__periph__gpio.html#ga48ba4c318486db320a4bf6237b49d61c).

You could [lock](https://doc.riot-os.org/group__core__sync__mutex.html#gac50891b4f3d2930d34c173567d739201) on a locked mutex here and unlock it in the interrupt callback - so this will only wait as long as it takes to get the measurement, not always the maximum time.

Reply:
From my understanding, it makes more sense to use mutex in places where the object is being changed in multiple threads and synchronization is required there. I am not sure if the overhead of adding the mutex implementation just to remove polling is worth it. What do you think ?

Reply:
Using a mutex for signalling is a common pattern. Just add a mutex to your `sdp3x_t` struct:

```C
typedef struct {
    sdp3x_params_t params;    /**< Device initialization parameters */
    mutex_t ready;
} sdp3x_t;
```

and initialize it in the locked state

```C
mutex_init(&dev->ready);
mutex_lock(&dev->ready);
```

Then, here instead of polling just try to lock the mutex again with `mutex_lock(&dev->ready);` - or use `xtimer_mutex_lock_timeout()` if there is a possibility that the interrupt never comes.

The simply `mutex_unlock(&dev->ready)`in the interrupt.

Mutex is used in many places in RIOT, so it's unlikely you'll pull any more code in.
And instead of always waiting the worst-case time, you are waiting exactly as long as it takes to get the measurement.

Reply:
I had the following two questions then :
1. The current code also waits only exactly for the amount of time it is required for the data to be ready (when the irq pin is actually connected and giving correct values). So, is the main concern that polling is not good code practice ?

This starts timer is the irq pin is not giving the desired value (which should be 256 if the sensor has started measuring data):
```
if (get_irq_pin_value(dev) == 0) {
        xtimer_usleep(46000);
    }
```
This waits for the irq_pin to change from 256 (measuring value) to 0 (data ready value):
```
 while (get_irq_pin_value(dev) != 0  && (xtimer_now_usec() - time_now) < 46000) {}
```
2. I planned on doing the way you mentioned above but the issue is that `sdp3x_t dev` is `const` everywhere in the code. Therefore, when I stored mutex inside the dev it creates errors since I am technically trying to change a const variable. So, what would you recommend in this case ?

Reply:
> The current code also waits only exactly for the amount of time it is required for the data to be ready

That is the *maximum* time.
If you are just waiting a fixed time, there would be no need for an interrupt pin, you could just read a status register instead to figure out if the measurement is ready.
If you waited a bit to short, you would then have to wait a full cycle again even if the measurement would be ready within a few µs already.

That's why there is an interrupt pin - you don't have to guess when the measurement is ready, the sensor will tell you.

> I planned on doing the way you mentioned above but the issue is that `sdp3x_t *dev` is const everywhere in the code.

`sdp3x_t *dev`  shouldn't be `const`. It makes sense for `sdp3x_params` to be `const`, but the driver handle itself doesn't have to be `const`.

Reply:
I just checked the data sheet and it says

> Measurement time Triggered mode Min: 40 Typ: 45 Max: 50 (ms)

So it should be enough to change `(xtimer_now_usec() - time_now) < 46000` to `(xtimer_now_usec() - time_now) < 50000` right?
That way we wait for the interrupt to happen with `get_irq_pin_value(dev) != 0` and stop if it occurs. If it did not occur during the 50ms specified in the data sheet we assume an error and stop waiting through `(xtimer_now_usec() - time_now) < 50000`.
Also we can probably remove this:
```C   
 /* If for some reason the pin values are not as expected */
     if (get_irq_pin_value(dev) == 0) {
         xtimer_usleep(DATA_READY_SLEEP);
     }
```
Of course we can also change it to use a mutex.

Reply:
> So it should be enough to change `(xtimer_now_usec() - time_now) < 46000` to `(xtimer_now_usec() - time_now) < 50000` right?

If you don't mind needlessly waiting 10ms extra :wink: 

But really, why not just use the interrupt pin as an interrupt source - the code will be much simpler.